### PR TITLE
Improving MLP

### PR DESCRIPTION
**Pull request by LAndersen1**
Addition:
-MLP-Cross
Improvement:
-MLP

### CODE REVIEW

File: bayesian_protein/surrogates.py [hunk 1]
```diff
@@ -376,13 +383,222 @@ def expected_improvement(self, x: np.ndarray, best_seen: float, smiles) -> float
...
+        loss_pred = torch.square(y_real-y_pred)/var_pred
+        #loss_var = weighting_var*torch.log(var_pred)
+        #loss_var = 10*torch.log(var_pred)**3
+        loss_var = torch.square(var_pred)
+        #loss_var = var_pred
+        losses = weighting_pred*loss_pred+weighting_var*loss_var
+        return torch.mean(losses,axis=0)
+
+    def __loss_fn(self, y_pred, y_real, var_pred):
+        weighting_pred = 0/2
```
Comment (maxrdu) [inline bayesian_protein/surrogates.py:455]: Why is this 0? In this case we are only weighting the variance loss terms. Is there any explanation for the weights?

File: bayesian_protein/surrogates.py [hunk 2]
```diff
@@ -376,13 +383,222 @@ def expected_improvement(self, x: np.ndarray, best_seen: float, smiles) -> float
...
+        var = []
+        self.model.eval()
+        for batch in loader:
+            batch = batch.to(self.device)
+            output = self.model.forward(batch).cpu().numpy().reshape((-1,2))
+
+            mean.append(np.atleast_1d(output[:,0]))
+            var.append(np.atleast_1d(np.sqrt(output[:,1])))
+            
+        return np.concatenate(mean), np.concatenate(var)
```
Comment (maxrdu) [inline bayesian_protein/surrogates.py]: Expected Improvement expects that we return the standard deviation here. We need to update the loss accordingly

File: bayesian_protein/surrogates.py [hunk 3]
```diff
@@ -525,6 +761,178 @@ def forward(self, X: np.ndarray, smiles) -> Tuple[np.ndarray, np.ndarray]:
...
+            self.models.append(m)
+        self.known_input_size = input_size
+
+    def _restore(self):
+        if self.models is None:
+            raise ValueError(
+                "Can only be called once 'update' has been called at least once."
+            )
+
+        self._build_model(self.known_input_size)
```
Comment (maxrdu) [inline bayesian_protein/surrogates.py:605]: This does not load the original weights for the model, but initializes new models. Not an issue but we should note this somewhere, because it's different to the other MLP version

File: bayesian_protein/surrogates.py [hunk 4]
```diff
@@ -525,6 +761,178 @@ def forward(self, X: np.ndarray, smiles) -> Tuple[np.ndarray, np.ndarray]:
...
+        blocks = [b1,b2,b3,b4,b5]
+        for i in range(len(self.models)):
+            train_ds_block = []
+            for q in range(len(self.models)):
+                if q==i:
+                    continue
+                train_ds_block.append(blocks[q])
+            train_ds = torch.utils.data.ConcatDataset(train_ds_block)
+            val_ds = blocks[i]
+            if len(train_ds) <= 0:
```
Comment (maxrdu) [inline bayesian_protein/surrogates.py]: Can this happen?

File: bayesian_protein/surrogates.py (hunk 2 above)
Comment (LAndersen1) [reply bayesian_protein/surrogates.py]: The square root is already computed (so it is already std), hence the variable naming might be misleading, but the values should be correct.

It is to note though that the variance is approximated and one could justify modifying this value especially in accordance with the loss function (since the loss function is what "gives" the variable the "meaning" to be a "variance")

File: bayesian_protein/surrogates.py (hunk 1 above)
Comment (LAndersen1) [reply bayesian_protein/surrogates.py:455]: This is an experimental loss function with the current parameters providing the best values. I found that putting relatively high weights on the variance term results in the best performance (the weight that is set to 0 resembles the standard term without any variance (there is a third term that combines the original formula with the variance))

File: bayesian_protein/surrogates.py (hunk 3 above)
Comment (LAndersen1) [reply bayesian_protein/surrogates.py:605]: Intended, since the model could be stuck in a weired local minimum from previous runs

File: bayesian_protein/surrogates.py (hunk 4 above)
Comment (LAndersen1) [reply bayesian_protein/surrogates.py]: In the first iteration, if there is only 1 sample provided the train_ds is 0 for 1 of the 5 blocks. Hence, I decided to use the same sample for the train and val ds.

File: bayesian_protein/surrogates.py (hunk 3 above)
Comment (maxrdu) [reply bayesian_protein/surrogates.py:605]: When referring to the original weights, I mean the variables before training in the very first iteration, i.e. the random initialization of the weights.

File: bayesian_protein/surrogates.py (hunk 4 above)
Comment (maxrdu) [reply bayesian_protein/surrogates.py]: Should be fixed with the latest [commit](https://github.com/uds-lsv/chemical-lm-active-learning/pull/2/commits/53edd12ef9b7b3d199a02bbad7aa1e1b43b76f15)
